### PR TITLE
feat: LT-990 Case SLA

### PR DESCRIPTION
**Pull request by BenAndrew-Kainos**
Creates the SLA Updater background task & Case fields

**New fields;**
- sla_days: Number of days since case was created
- sla_remaining_days: Number of days left to process to meet SLA goal
- last_closed_at: Last time case was closed. Needed becuase previously closed cases don't update SLA
- sla_updated_at: Last time SLA was recalculated. used to ensure we don't recalculate twice a day

**New key functions;**
- update_cases_sla: Background task run once a day to update cases SLA.
- get_application_target_sla: Get the goal days for a given application type
- get_bank_holidays: Use the GOV API to get bank holidays & save a local backup

### CODE REVIEW

File: cases/models.py [hunk 1]
```diff
@@ -59,6 +59,11 @@ class Case(TimestampableModel):
     )
     case_officer = models.ForeignKey(GovUser, null=True, on_delete=models.DO_NOTHING)
     copy_of = models.ForeignKey("self", default=None, null=True, on_delete=models.DO_NOTHING)
+    last_closed_at = models.DateTimeField(blank=True, null=True)
+
+    sla_days = models.IntegerField(null=False, blank=False, default=0)
```
Comment (Mark-Spiers-kainos) [inline cases/models.py]: These are very big fields in terms of length. They also include negative numbers. Perhaps look at SmallIntegerField/PositiveSmallIntegerField
https://docs.djangoproject.com/en/3.0/ref/models/fields/#positivesmallintegerfield
https://docs.djangoproject.com/en/3.0/ref/models/fields/#smallintegerfield

Comment (BenAndrew-Kainos) [reply cases/models.py]: Great suggestion!

File: cases/models.py [hunk 2]
```diff
@@ -59,6 +59,11 @@ class Case(TimestampableModel):
     )
     case_officer = models.ForeignKey(GovUser, null=True, on_delete=models.DO_NOTHING)
     copy_of = models.ForeignKey("self", default=None, null=True, on_delete=models.DO_NOTHING)
+    last_closed_at = models.DateTimeField(blank=True, null=True)
+
+    sla_days = models.PositiveSmallIntegerField(null=False, blank=False, default=0)
+    sla_remaining_days = models.SmallIntegerField(null=True, blank=True)
+    sla_updated_at = models.DateTimeField(blank=True, null=True)
```
Comment (stephanieboyd) [inline cases/models.py]: Is `blank=True` necessary?

File: cases/sla.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+from cases.enums import CaseTypeSubTypeEnum
+from cases.models import Case
+
+SLA_UPDATE_TASK_TIME = time(0, 0, 0)
+SLA_UPDATE_CUTOFF_TIME = time(18, 0, 0)
+BANK_HOLIDAY_API = "https://www.gov.uk/bank-holidays.json"
+BACKUP_FILE_NAME = "bank-holidays.csv"
+
+STANDARD_APPLICATION_TARGET = 20
```
Comment (stephanieboyd) [inline cases/sla.py]: It could be clearer that this is a target in days

File: cases/sla.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+def get_bank_holidays():
+    """
+    Uses the GOV bank holidays API.
+    If it can connect to the API, it extracts the list of bank holidays,
+    saves a backup of this list as a CSV and returns the list.
+    If it cannot connect to the service it will use the CSV backup and returns the list.
+    """
+    data = []
+    r = requests.get(BANK_HOLIDAY_API)
+    if r.status_code != 200:
```
Comment (stephanieboyd) [inline cases/sla.py]: HTTP status, please

File: cases/sla.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                        sla_days=F("sla_days") + 1, sla_remaining_days=F("sla_remaining_days") - 1, sla_updated_at=date
+                    )
+                )
+                logging.info(f"SLA Update Successful: Updated {results} cases")
+                return results
+        except Exception as e:  # noqa
+            logging.error(e)
+            return False
+
+    logging.info("SLA Update Ignored: Non-working day")
```
Comment (stephanieboyd) [inline cases/sla.py]: Super minor - not a fan that the message is  `ignored`. Perhaps `not performed`

File: cases/models.py (hunk 2 above)
Comment (BenAndrew-Kainos) [reply cases/models.py]: I don't think so

File: cases/sla.py (hunk 3 above)
Comment (BenAndrew-Kainos) [reply cases/sla.py]: Done

File: cases/sla.py (hunk 4 above)
Comment (BenAndrew-Kainos) [reply cases/sla.py]: Done

File: cases/sla.py (hunk 5 above)
Comment (BenAndrew-Kainos) [reply cases/sla.py]: Done

**Pull request merged**